q2_composition/_ancom.py

```diff
@@ -5,3 +5,73 @@
 #
 # The full license is in the file LICENSE, distributed with this software.
 # ----------------------------------------------------------------------------
+import tempfile
+import pandas as pd
+import numpy as np
+import os
+import qiime2
+
+
+def ancombc(table: pd.DataFrame, metadata: qiime2.Metadata, formula: str,
+            p_adj_method: str = "holm", zero_cut: float = 0.90,
+            lib_cut: int = 1000, group: str = None, struc_zero: bool = True,
+            neg_lb: bool = True, tol: float = 1e-5, max_iter: int = 100,
+            conserve: bool = True, alpha: float = 0.05) -> pd.DataFrame:
+
+    # create series from metadata column
+    meta = metadata.to_dataframe()
+
+    # check for variable lengths and warn if there's only one value per group.
+    # ANCOM will fail silently later bc of this and that is harder to debug.
+    variables = np.unique(np.hstack([x.split("*")
+                                    for x in formula.split("+")]))
+    variables = np.array([x.strip() for x in variables])
+    var_counts = pd.DataFrame.from_dict(orient='index', data={
+        var: {'n_groups': len(meta[var].dropna().unique())}
+        for var in variables
+        })
+    if (var_counts['n_groups'] < 2).all():
+        raise ValueError("None of the columns in the metadata satisfy"
+                         " ANCOM-BC's requirements. All columns in the"
+                         " formula should have more than one value.")
+
+    # filter the metadata so only the samples present in the table are used.
+    # this also re-orders it for the correct condition selection.
+    # it must be re-ordered for ANCOM-BC to correctly input the conditions.
```

Comment:
Shouldn't need to do this at all because the phyloseq object should only be using metadata that's associated w/the table - but need to confirm

Reply:
Confirmed that metadata is unchanged before/after `meta.loc[list(table.index)]` so will remove this in next commit.

![Screen Shot 2022-06-29 at 3 39 57 PM](https://user-images.githubusercontent.com/54517601/176557701-e9f3c571-7399-4492-83cc-baad72c2aee9.png)
![Screen Shot 2022-06-29 at 3 37 02 PM](https://user-images.githubusercontent.com/54517601/176557418-bbdcdf26-2ba9-4ad4-9fce-0fea741e0150.png)

![Screen Shot 2022-06-29 at 3 40 06 PM](https://user-images.githubusercontent.com/54517601/176557702-6574c927-dd6d-4e2b-913d-73b15a25d647.png)
![Screen Shot 2022-06-29 at 3 37 15 PM](https://user-images.githubusercontent.com/54517601/176557420-da8ac7b6-5f1b-477b-a0ef-03efecba0cd1.png)

Reply:
I don't think this test proves anything yet because you aren't looking at the table after creating the phyloseq object which would happen in R. That's where a difference may or may not be observable.

With this particular example data, it turns out the table is exactly matched, so it doesn't work very well as a test.

Reply:
circling back on this for posterity - confirmed that phyloseq only takes an intersection of the table IDs and metadata IDs. error on IDs not present in metadata has been added as validation before the phyloseq object is created.